### PR TITLE
fix: restrict file permissions on auth, config, and log files to prevent local credential theft

### LINKED ISSUE #246: [BUG] Auth, config, context, and log files created with world-readable permissions on Linux/macOS
### Bug Description

Auth files (`auth.json`, `config.json`, `context-{pid}.json`) and debug log directories are created with default OS permissions (`0o644` for files, `0o755` for directories), making them world-readable on multi-user Linux/macOS systems. This exposes MSAL token caches, auth metadata, and debug logs containing sensitive API responses to other local users.

**CWE-276: Incorrect Default Permissions**

| File | Data at Risk | Current Permissions |
|------|-------------|---------------------|
| `~/.config/fab/` | Container for all auth/config | `0o755` (world-traversable) |
| `~/.config/fab/auth.json` | Tenant ID, SPN client ID, identity type | `0o644` (world-readable) |
| `~/.config/fab/config.json` | Azure subscription IDs, capacity settings | `0o644` (world-readable) |
| `~/.config/fab/context-{pid}.json` | Navigation state (workspace/item names) | `0o644` (world-readable) |
| Log directory | Parent of debug log with HTTP traffic | `0o755` (world-traversable) |

### fabric-cli version

1.6.1

### Python version

Python 3.12

### Operating System

Linux

### Steps to Reproduce

1. Install fabric-cli on a multi-user Linux/macOS system
2. Run `fab login` or any command that writes auth/config files
3. Check permissions: `ls -la ~/.config/fab/`
4. Observe files are `0o644` (world-readable) and directories are `0o755` (world-traversable)

### Expected Behavior

Auth, config, context, and log files should be created with restrictive permissions (`0o600` for files, `0o700` for directories) so only the owning user can read them.

### Actual Behavior

Files are created with default OS permissions, allowing any local user to read sensitive credentials and configuration data.

### Possible Solution

Use `os.open()` with explicit mode `0o600` instead of `open()` for writing sensitive files. Create directories with `mode=0o700`.

### PR DESCRIPTION
**Pull request by iemejia**
- Resolves #246

## Summary

Auth files (`auth.json`, `config.json`, `context-{pid}.json`) and debug log directories were created with default OS permissions (`0o644` for files, `0o755` for directories), making them world-readable on multi-user Linux/macOS systems. This exposed MSAL token caches, auth metadata, and debug logs containing sensitive API responses to other local users.

## Security Impact

| File | Data at Risk | Old Permissions | New Permissions |
|------|-------------|-----------------|-----------------| 
| `~/.config/fab/` | Container for all auth/config | `0o755` (world-traversable) | `0o700` (owner-only) |
| `~/.config/fab/auth.json` | Tenant ID, SPN client ID, identity type | `0o644` (world-readable) | `0o600` (owner-only) |
| `~/.config/fab/config.json` | Azure subscription IDs, capacity settings | `0o644` (world-readable) | `0o600` (owner-only) |
| `~/.config/fab/context-{pid}.json` | Navigation state (workspace/item names) | `0o644` (world-readable) | `0o600` (owner-only) |
| Log directory | Parent of debug log with HTTP traffic | `0o755` (world-traversable) | `0o700` (owner-only) |

**CWE-276: Incorrect Default Permissions**

## Changes

- **`fab_auth.py`**: Write `auth.json` with `0o600` via `os.open()` instead of `open()`
- **`fab_state_config.py`**: Create `~/.config/fab/` with `mode=0o700`; write `config.json` with `0o600` via `os.open()` with a new `_write_restricted_file()` helper
- **`fab_logger.py`**: Create log directory with `mode=0o700`
- **`fab_context.py`**: Write `context-{pid}.json` with `0o600` via `os.open()`

## Cross-Platform Behavior

- **Linux/macOS**: Fixes enforce `0o600`/`0o700` correctly — resolves the vulnerability
- **Windows**: The `mode` parameter is a no-op (Windows uses ACLs, not POSIX permissions). Default Windows ACLs on user profile directories already restrict access to the owner, so the original behavior was not vulnerable there. These changes are safely a no-op on Windows.

## Tests

6 new tests verifying:
- Config directory created with `0o700`
- `config.json` written with `0o600` (initial write and overwrite)
- `auth.json` written with `0o600` (initial write and overwrite)
- Log directory created with `0o700`

All existing tests pass (488 total).

### CODE REVIEW

File: src/fabric_cli/core/fab_state_config.py [hunk 1]
```diff
@@ -29,8 +29,19 @@ def read_config(file_path) -> dict:
...
+def _write_restricted_file(file_path, content):
+    """Write content to a file with owner-only permissions (0o600)."""
+    fd = os.open(file_path, os.O_WRONLY | os.O_CREAT | os.O_TRUNC, 0o600)
+    try:
+        with os.fdopen(fd, "w") as file:
+            file.write(content)
+    except BaseException:
+        # fd is already closed by os.fdopen on success; only close on failure
+        # before os.fdopen wraps it
+        raise
```
[the remark this answers is not in the document]
Comment (iemejia) [reply src/fabric_cli/core/fab_state_config.py]: Fixed. `_write_restricted_file()` now catches `Exception` (not `BaseException`) and closes `fd` in the except block before re-raising. Also added `_chmod_if_posix()` after write to tighten pre-existing files.

File: src/fabric_cli/core/fab_state_config.py [hunk 2]
```diff
@@ -29,8 +29,19 @@ def read_config(file_path) -> dict:
 
 
 def write_config(data):
-    with open(config_file, "w") as file:
-        json.dump(data, file, indent=4)
+    _write_restricted_file(config_file, json.dumps(data, indent=4))
```
[the remark this answers is not in the document]
Comment (iemejia) [reply src/fabric_cli/core/fab_state_config.py]: Fixed. Added `_chmod_if_posix(file_path, 0o600)` after every write to enforce permissions on pre-existing files from older CLI versions. Also added a dedicated test (`test_write_config_tightens_permissions_on_existing_file`) that creates a file with `0o644` and verifies it gets tightened to `0o600` after overwrite.

File: src/fabric_cli/core/fab_auth.py [hunk 3]
```diff
@@ -56,7 +56,8 @@ def __init__(self):
         self._load_env()
 
     def _save_auth(self):
-        with open(self.auth_file, "w") as file:
+        fd = os.open(self.auth_file, os.O_WRONLY | os.O_CREAT | os.O_TRUNC, 0o600)
+        with os.fdopen(fd, "w") as file:
             file.write(json.dumps(self._get_auth_info()))
```
[the remark this answers is not in the document]
Comment (iemejia) [reply src/fabric_cli/core/fab_auth.py]: Fixed. `_save_auth()` now delegates to the shared `_write_restricted_file()` helper from `fab_state_config`, which handles fd-leak safety and chmod for pre-existing files. Added `test_save_auth_tightens_permissions_on_existing_file` to verify.

File: src/fabric_cli/core/fab_context.py [hunk 4]
```diff
@@ -159,7 +159,12 @@ def _save_context_to_file(self) -> None:
...
 
             try:
-                with open(self._context_file, "w") as f:
+                fd = os.open(
+                    self._context_file,
+                    os.O_WRONLY | os.O_CREAT | os.O_TRUNC,
+                    0o600,
+                )
+                with os.fdopen(fd, "w") as f:
                     json.dump(context_data, f)
```
[the remark this answers is not in the document]
Comment (iemejia) [reply src/fabric_cli/core/fab_context.py]: Fixed. Context file save now delegates to the shared `_write_restricted_file()` helper, getting both fd-leak protection and chmod for pre-existing files. Updated the context persistence tests to mock the new code path.

File: src/fabric_cli/core/fab_state_config.py [hunk 5]
```diff
@@ -11,7 +11,7 @@
 def config_location():
     _location = expanduser("~/.config/fab/")
     if not exists(_location):
-        os.makedirs(_location)
+        os.makedirs(_location, mode=0o700)
     return _location
```
[the remark this answers is not in the document]
Comment (iemejia) [reply src/fabric_cli/core/fab_state_config.py:18]: Fixed. Changed to `os.makedirs(_location, mode=0o700, exist_ok=True)` and added `_chmod_if_posix(_location, 0o700)` to tighten pre-existing directories. Added `test_config_location_tightens_permissions_on_existing_directory` that creates a `0o755` dir and verifies it becomes `0o700`.

File: tests/test_core/test_fab_state_config.py [hunk 6]
```diff
@@ -165,3 +165,65 @@ def test_init_defaults_preserves_user_overrides_success(monkeypatch, tmp_path):
...
+    monkeypatch, tmp_path
+):
+    """Verify config directory is created with mode 0o700 (owner-only access)."""
+    config_dir = tmp_path / "fab_config_test"
+    monkeypatch.setattr(
+        cfg, "config_location", lambda: _create_restricted_dir(str(config_dir))
+    )
+    # Call our helper which mimics the real config_location logic
+    location = _create_restricted_dir(str(config_dir))
+    assert os.path.isdir(location)
```
[the remark this answers is not in the document]
Comment (iemejia) [reply tests/test_core/test_fab_state_config.py]: Fixed. Test now patches `expanduser` to redirect into `tmp_path` and calls the real `cfg.config_location()`. Removed unused `_create_restricted_dir` helper and `config_dir` variable. Added `@pytest.mark.skipif(os.name == 'nt')` to all permission assertion tests.

File: src/fabric_cli/core/fab_logger.py [hunk 7]
```diff
@@ -156,7 +156,7 @@ def log_debug_http_request_exception(e):
 def _get_log_file_path():
     """Create a log file path in the user's log directory."""
     log_dir = user_log_dir("fabric-cli")
-    os.makedirs(log_dir, exist_ok=True)
+    os.makedirs(log_dir, mode=0o700, exist_ok=True)
     return os.path.join(log_dir, "fabcli_debug.log")
```
[the remark this answers is not in the document]
Comment (iemejia) [reply src/fabric_cli/core/fab_logger.py:165]: Fixed. Added `_chmod_if_posix(log_dir, 0o700)` after `os.makedirs` in `_get_log_file_path()`, matching the pattern used in `config_location()`.

File: src/fabric_cli/core/fab_state_config.py [hunk 8]
```diff
@@ -3,15 +3,25 @@
...
 from fabric_cli.core import fab_constant
 
 
+def _chmod_if_posix(path, mode):
+    """Best-effort chmod; no-op on Windows where POSIX modes don't apply."""
+    if os.name != "nt":
+        try:
+            os.chmod(path, mode)
+        except OSError:
+            pass
```
[the remark this answers is not in the document]
Comment (iemejia) [reply src/fabric_cli/core/fab_state_config.py]: Fixed. `_chmod_if_posix()` now emits a `logging.warning()` when chmod fails on POSIX, so operators can detect restrictive filesystems or unexpected ownership conditions.

File: .changes/unreleased/fixed-20260607-134043.yaml [hunk 9]
```diff
@@ -0,0 +1,6 @@
+kind: fixed
+body: Restrict file permissions on auth, config, and log files to prevent local credential theft on multi-user systems
+time: 2026-06-07T13:40:43+02:00
+custom:
+    Author: iemejia
+    AuthorLink: https://github.com/iemejia
```
[the remark this answers is not in the document]
Comment (iemejia) [reply .changes/unreleased/fixed-20260607-134043.yaml:6]: Fixed. Updated the changelog text to accurately describe the scope: "Restrict file and directory permissions on auth, config, context, and log paths".

File: tests/test_core/test_fab_state_config.py [hunk 10]
```diff
@@ -3,8 +3,11 @@
 
 import json
 import os
+import sys
```
[the remark this answers is not in the document]
Comment (iemejia) [reply tests/test_core/test_fab_state_config.py]: Fixed. Removed unused `sys` import.

File: tests/test_core/test_fab_state_config.py [hunk 11]
```diff
@@ -165,3 +168,86 @@ def test_init_defaults_preserves_user_overrides_success(monkeypatch, tmp_path):
...
+    os.name == "nt", reason="POSIX permission tests not applicable on Windows"
+)
+
+
+@_skip_on_windows
+def test_config_location_creates_directory_with_restricted_permissions(
+    monkeypatch, tmp_path
+):
+    """Verify config directory is created with mode 0o700 (owner-only access)."""
+    config_dir = str(tmp_path / ".config" / "fab")
```
[the remark this answers is not in the document]
Comment (iemejia) [reply tests/test_core/test_fab_state_config.py]: Fixed. Removed unused `config_dir` variable.

File: src/fabric_cli/core/fab_state_config.py [hunk 12]
```diff
@@ -29,8 +44,29 @@ def read_config(file_path) -> dict:
 
 
 def write_config(data):
-    with open(config_file, "w") as file:
-        json.dump(data, file, indent=4)
+    _write_restricted_file(config_file, json.dumps(data, indent=4))
+
+
+def _write_restricted_file(file_path, content):
```
Comment (aviatco) [inline src/fabric_cli/core/fab_state_config.py]: Please extract this to utils folder and reused it in other files as well.

File: src/fabric_cli/core/fab_auth.py [hunk 13]
```diff
@@ -56,8 +56,9 @@ def __init__(self):
         self._load_env()
 
     def _save_auth(self):
-        with open(self.auth_file, "w") as file:
-            file.write(json.dumps(self._get_auth_info()))
+        from fabric_cli.core.fab_state_config import _write_restricted_file
```
Comment (aviatco) [inline src/fabric_cli/core/fab_auth.py]: I recommend creating a helper under the utils folder that handles file creation, opening, and writing.

This logic should not be part of fab_state_config, and classes should not depend on it just for this functionality.

File: src/fabric_cli/core/fab_context.py [hunk 14]
```diff
@@ -159,8 +159,11 @@ def _save_context_to_file(self) -> None:
             }
 
             try:
-                with open(self._context_file, "w") as f:
-                    json.dump(context_data, f)
+                from fabric_cli.core.fab_state_config import _write_restricted_file
+
+                _write_restricted_file(
```
Comment (aviatco) [inline src/fabric_cli/core/fab_context.py]: Same comment as above

File: src/fabric_cli/core/fab_logger.py [hunk 15]
```diff
@@ -156,10 +156,21 @@ def log_debug_http_request_exception(e):
 def _get_log_file_path():
     """Create a log file path in the user's log directory."""
     log_dir = user_log_dir("fabric-cli")
-    os.makedirs(log_dir, exist_ok=True)
+    os.makedirs(log_dir, mode=0o700, exist_ok=True)
```
Comment (aviatco) [inline src/fabric_cli/core/fab_logger.py]: please move this logic to the helper and reuse it

File: src/fabric_cli/core/fab_logger.py [hunk 16]
```diff
@@ -156,10 +156,21 @@ def log_debug_http_request_exception(e):
...
+    # Enforce permissions on pre-existing directories from older versions
+    _chmod_if_posix(log_dir, 0o700)
```
Comment (aviatco) [inline src/fabric_cli/core/fab_logger.py]: Also here, please move it to the helper and reuse it

File: src/fabric_cli/core/fab_logger.py [hunk 17]
```diff
@@ -156,10 +156,21 @@ def log_debug_http_request_exception(e):
...
     return os.path.join(log_dir, "fabcli_debug.log")
 
 
+def _chmod_if_posix(path, mode):
+    """Best-effort chmod with warning on failure; no-op on Windows."""
+    if os.name != "nt":
+        try:
+            os.chmod(path, mode)
+        except OSError:
+            pass
```
Comment (aviatco) [inline src/fabric_cli/core/fab_logger.py]: Missing warning on failure

File: src/fabric_cli/core/fab_logger.py [hunk 18]
```diff
@@ -156,10 +156,21 @@ def log_debug_http_request_exception(e):
...
-    os.makedirs(log_dir, exist_ok=True)
+    os.makedirs(log_dir, mode=0o700, exist_ok=True)
+    # Enforce permissions on pre-existing directories from older versions
+    _chmod_if_posix(log_dir, 0o700)
     return os.path.join(log_dir, "fabcli_debug.log")
 
 
+def _chmod_if_posix(path, mode):
+    """Best-effort chmod with warning on failure; no-op on Windows."""
+    if os.name != "nt":
```
Comment (aviatco) [inline src/fabric_cli/core/fab_logger.py]: Consider extract it to global variable like _POSIX or similar and add comment explaining what it is

File: src/fabric_cli/core/fab_logger.py
Comment (aviatco) [inline src/fabric_cli/core/fab_logger.py]: There is log file rotation in place, so this needs to be handled as well, ensuring that every file, whether created, opened, or re-opened during rotation, is assigned the correct permissions.

File: src/fabric_cli/core/fab_state_config.py (hunk 12 above)
Comment (iemejia) [reply src/fabric_cli/core/fab_state_config.py]: Done. Extracted `_write_restricted_file` (now `write_restricted_file`) and `_chmod_if_posix` (now `chmod_if_posix`) along with `create_restricted_dir` to a new `src/fabric_cli/utils/fab_file_permissions.py` module. All callers now import from there.

File: src/fabric_cli/core/fab_auth.py (hunk 13 above)
Comment (iemejia) [reply src/fabric_cli/core/fab_auth.py]: Done. `_save_auth()` now imports `write_restricted_file` directly from `fabric_cli.utils.fab_file_permissions` — no dependency on `fab_state_config` for this functionality.

File: src/fabric_cli/core/fab_context.py (hunk 14 above)
Comment (iemejia) [reply src/fabric_cli/core/fab_context.py]: Done. Context file save now imports `write_restricted_file` from `fabric_cli.utils.fab_file_permissions`.

File: src/fabric_cli/core/fab_logger.py (hunk 15 above)
Comment (iemejia) [reply src/fabric_cli/core/fab_logger.py]: Done. Log directory creation now calls `create_restricted_dir(log_dir)` from the shared utils module.

File: src/fabric_cli/core/fab_logger.py (hunk 16 above)
Comment (iemejia) [reply src/fabric_cli/core/fab_logger.py]: Done. The local `_chmod_if_posix` in `fab_logger` is removed; all permission enforcement uses `chmod_if_posix` from `fabric_cli.utils.fab_file_permissions`.

File: src/fabric_cli/core/fab_logger.py (hunk 17 above)
Comment (iemejia) [reply src/fabric_cli/core/fab_logger.py]: Fixed. The shared `chmod_if_posix` in the new utils module emits a `logging.warning()` on POSIX when chmod fails, which was missing in the previous `fab_logger` local copy.

File: src/fabric_cli/core/fab_logger.py (hunk 18 above)
Comment (iemejia) [reply src/fabric_cli/core/fab_logger.py]: Done. Added `_IS_POSIX = os.name != "nt"` as a module-level constant in `fab_file_permissions.py` with a comment: "True on Linux/macOS; False on Windows where POSIX permission bits are a no-op." Used across both `chmod_if_posix` and the log file handler.

File: src/fabric_cli/core/fab_logger.py
Comment (iemejia) [reply src/fabric_cli/core/fab_logger.py]: Done. Introduced `_RestrictedRotatingFileHandler` subclass of `RotatingFileHandler` that:
- Overrides `_open()` with a custom opener to create files with `0o600`
- Overrides `doRollover()` to enforce `0o600` on all rotated backup files
- Calls `chmod_if_posix` on the initial log file to tighten pre-existing files

This ensures every file created, opened, or re-opened during rotation gets the correct permissions.

File: src/fabric_cli/utils/fab_file_permissions.py [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+    """Best-effort chmod with warning on failure; no-op on Windows.
+
+    Emits a warning when chmod fails on POSIX so operators can detect
+    restrictive filesystems or unexpected ownership/ACL conditions.
+    """
+    if _IS_POSIX:
+        try:
+            os.chmod(path, mode)
+        except OSError as e:
+            _logger.warning("Failed to set permissions %o on %s: %s", mode, path, e)
```
[the remark this answers is not in the document]
Comment (iemejia) [reply src/fabric_cli/utils/fab_file_permissions.py]: The best-effort approach is intentional. Failing hard on chmod would break the CLI on systems with restrictive filesystems (certain container volumes, FUSE mounts, or NFS exports where chmod is a no-op). The warning allows operators to detect the condition in logs. The actual credential protection comes primarily from the `os.open(..., 0o600)` at creation time — the chmod is a defense-in-depth for upgrading pre-existing files from older CLI versions.

File: src/fabric_cli/core/fab_logger.py [hunk 20]
```diff
@@ -13,6 +13,11 @@
 import fabric_cli.core.fab_constant as fab_constant
 import fabric_cli.core.fab_state_config as fab_state_config
 import fabric_cli.utils.fab_ui as utils_ui
+from fabric_cli.utils.fab_file_permissions import (
+    _IS_POSIX,
+    chmod_if_posix,
+    create_restricted_dir,
+)
```
[the remark this answers is not in the document]
Comment (iemejia) [reply src/fabric_cli/core/fab_logger.py]: `_IS_POSIX` is intentionally underscore-prefixed as it's an internal implementation detail shared between sibling modules within the same package (`fabric_cli`), not part of the public API surface. Importing private symbols across modules within the same package is a common Python pattern for internal constants that shouldn't be relied upon by external consumers.

File: src/fabric_cli/utils/fab_file_permissions.py [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+def write_restricted_file(file_path: str, content: str) -> None:
+    """Write content to a file with owner-only permissions (0o600).
+
+    Handles both new file creation and tightening permissions on
+    pre-existing files from older CLI versions.
+    """
+    fd = os.open(file_path, os.O_WRONLY | os.O_CREAT | os.O_TRUNC, 0o600)
+    try:
+        with os.fdopen(fd, "w") as file:
+            file.write(content)
```
[the remark this answers is not in the document]
Comment (iemejia) [reply src/fabric_cli/utils/fab_file_permissions.py]: Fixed. Added `encoding="utf-8"` to `os.fdopen()` for consistent JSON persistence across platforms/locales.

File: tests/test_core/test_fab_state_config.py [hunk 22]
```diff
@@ -148,20 +158,109 @@ def test_init_defaults_applies_missing_defaults_success(monkeypatch, tmp_path):
...
+    monkeypatch.setattr(
+        "fabric_cli.core.fab_state_config.expanduser",
+        lambda path: path.replace("~", str(tmp_path)),
+    )
+
+    location = cfg.config_location()
+    assert os.path.isdir(location)
+
+    mode = oct(os.stat(location).st_mode & 0o777)
+    assert mode == "0o700", f"Config directory has mode {mode}, expected 0o700"
```
[the remark this answers is not in the document]
Comment (iemejia) [reply tests/test_core/test_fab_state_config.py:206]: The tests already skip on Windows (`@pytest.mark.skipif(os.name == "nt")`). On POSIX systems where chmod doesn't work (e.g., some FUSE/network mounts), this is a rare edge case unlikely to affect CI. The trade-off of having strict permission assertions in tests is worth it to catch regressions — if a CI environment truly cannot chmod, those tests would need environment-specific skipping, which would be addressed if/when it occurs.

File: src/fabric_cli/core/fab_auth.py [hunk 23]
```diff
@@ -434,18 +434,18 @@ def _is_token_defined(self, scope):
     def acquire_token(self, scope: list[str], interactive_renew=True) -> dict:
         """
         Core MSAL token acquisition method that returns the full MSAL result.
-        
+
```
Comment (aviatco) [inline src/fabric_cli/core/fab_auth.py:437]: Please remove the spaces

File: src/fabric_cli/utils/fab_file_permissions.py (hunk 19 above)
Comment (aviatco) [inline src/fabric_cli/utils/fab_file_permissions.py]: Consider change waring to debug since in case of "permission demined" user has nothing to do we it. so, this is more debug error.

File: tests/test_core/test_fab_logger.py [hunk 24]
```diff
@@ -249,3 +249,24 @@ def mock_log_warning():
...
         yield mock
+
+
+# ── Security: log directory permissions ──────────────────────────────────────
+
+
+@pytest.mark.skipif(
+    os.name == "nt", reason="POSIX permission tests not applicable on Windows"
+)
+def test_get_log_file_path_creates_directory_with_restricted_permissions(
```
Comment (aviatco) [inline tests/test_core/test_fab_logger.py]: Please add _success/_failure suffix for the test's name

File: src/fabric_cli/utils/fab_secure_io.py [hunk 25]
```diff
@@ -0,0 +1,66 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
```
Comment (aviatco) [inline src/fabric_cli/utils/fab_secure_io.py:3]: Consider rename the file to clearer name such as fab_secure_io or something similar.

File: tests/test_core/test_fab_auth.py [hunk 26]
```diff
@@ -1064,3 +1068,54 @@ def test_auth_mode_migration(tmp_path):
...
+
+# region security: auth file permission tests
+
+_skip_on_windows = pytest.mark.skipif(
+    os.name == "nt", reason="POSIX permission tests not applicable on Windows"
+)
+
+
+@_skip_on_windows
+def test_save_auth_creates_file_with_restricted_permissions(tmp_path):
```
Comment (aviatco) [inline tests/test_core/test_fab_auth.py]: Please add _succses/_failure suffix

File: tests/test_core/test_fab_auth.py [hunk 27]
```diff
@@ -1064,3 +1068,54 @@ def test_auth_mode_migration(tmp_path):
...
+    assert data[con.IDENTITY_TYPE] == "user"
+
+
+@_skip_on_windows
+def test_save_auth_tightens_permissions_on_existing_file(tmp_path):
+    """Verify _save_auth() enforces 0o600 on a pre-existing permissive file."""
+    auth = FabAuth()
+    auth.auth_file = os.path.join(str(tmp_path), "auth.json")
+
+    # Create file with overly permissive mode (simulating old CLI version)
```
Comment (aviatco) [inline tests/test_core/test_fab_auth.py]: By "simulating an old CLI version," do you mean using world‑readable permissions? If so, it would be clearer to state this explicitly, as "old CLI version" is quite ambiguous.

File: tests/test_core/test_fab_logger.py [hunk 28]
```diff
@@ -249,3 +249,24 @@ def mock_log_warning():
...
+    monkeypatch, tmp_path
```
Comment (aviatco) [inline tests/test_core/test_fab_logger.py:263]: please test creating file verify the permission then open it and see user still can view it.
Also, did you verify file rotation is working as expected?

File: tests/test_core/test_fab_auth.py (hunk 26 above)
Comment (aviatco) [reply tests/test_core/test_fab_auth.py]: please fix it in other files as well

File: src/fabric_cli/core/fab_auth.py (hunk 23 above)
Comment (iemejia) [reply src/fabric_cli/core/fab_auth.py:437]: These are trailing-whitespace removals in docstrings, applied automatically by `black` (the project's required formatter). Since running `black` is a prerequisite before committing, these whitespace cleanups get included whenever we format the changed file. They are harmless but we understand they add noise to the diff.

File: src/fabric_cli/utils/fab_file_permissions.py (hunk 19 above)
Comment (iemejia) [reply src/fabric_cli/utils/fab_file_permissions.py]: Done. Changed `_logger.warning(...)` to `_logger.debug(...)` since the user cannot act on a permission-denied failure from chmod.

File: tests/test_core/test_fab_logger.py (hunk 24 above)
Comment (iemejia) [reply tests/test_core/test_fab_logger.py]: Done. Renamed to `test_get_log_file_path_creates_directory_with_restricted_permissions_success`. Also added two new tests: `test_log_file_created_with_restricted_permissions_success` (verifies log file is created with 0o600 and is readable by owner) and `test_log_file_rotation_preserves_restricted_permissions_success` (verifies rotated files maintain 0o600).

File: src/fabric_cli/utils/fab_secure_io.py (hunk 25 above)
Comment (iemejia) [reply src/fabric_cli/utils/fab_secure_io.py:3]: Done. Renamed `fab_file_permissions.py` to `fab_secure_io.py` and updated all imports across the codebase.

File: tests/test_core/test_fab_auth.py (hunk 26 above)
Comment (iemejia) [reply tests/test_core/test_fab_auth.py]: Done. Renamed all permission test functions to include `_success` suffix across `test_fab_auth.py`, `test_fab_logger.py`, and `test_fab_state_config.py`.

File: tests/test_core/test_fab_auth.py (hunk 27 above)
Comment (iemejia) [reply tests/test_core/test_fab_auth.py]: Good point. Clarified the comment to: `# Create file with world-readable permissions (0o644) to simulate a pre-existing file written by an older CLI version without hardening`. Applied the same fix in `test_fab_state_config.py`.

File: tests/test_core/test_fab_logger.py (hunk 28 above)
Comment (iemejia) [reply tests/test_core/test_fab_logger.py:263]: Done. Added `test_log_file_created_with_restricted_permissions_success` which creates a log file, verifies 0o600 permissions, writes a log entry, and reads it back to confirm the owner can still access it. Also added `test_log_file_rotation_preserves_restricted_permissions_success` which triggers log rotation with a small `maxBytes` and verifies both the base and rotated files have 0o600 permissions.

File: tests/test_core/test_fab_auth.py (hunk 26 above)
Comment (iemejia) [reply tests/test_core/test_fab_auth.py]: Done. All permission test names in `test_fab_auth.py`, `test_fab_logger.py`, and `test_fab_state_config.py` now include the `_success` suffix.

File: src/fabric_cli/utils/fab_secure_io.py [hunk 29]
```diff
@@ -0,0 +1,66 @@
...
+    except Exception:
+        # os.fdopen may fail before wrapping fd; close to avoid leak.
+        # If os.fdopen succeeded, fd is already closed by the with block.
+        try:
+            os.close(fd)
+        except OSError:
+            pass
+        raise
+    # Enforce permissions on pre-existing files from older versions
+    chmod_if_posix(file_path, 0o600)
```
[the remark this answers is not in the document]
Comment (iemejia) [reply src/fabric_cli/utils/fab_secure_io.py]: Fixed. `write_restricted_file()` now calls `chmod_if_posix(file_path, 0o600)` *before* truncation when the file already exists, so sensitive content is never written through a permissive fd. The `os.open(..., 0o600)` still handles the new-file case atomically.

File: src/fabric_cli/core/fab_logger.py [hunk 30]
```diff
@@ -13,6 +13,11 @@
 import fabric_cli.core.fab_constant as fab_constant
 import fabric_cli.core.fab_state_config as fab_state_config
 import fabric_cli.utils.fab_ui as utils_ui
+from fabric_cli.utils.fab_secure_io import (
+    _IS_POSIX,
+    chmod_if_posix,
+    create_restricted_dir,
+)
```
[the remark this answers is not in the document]
Comment (iemejia) [reply src/fabric_cli/core/fab_logger.py:20]: Fixed. Renamed `_IS_POSIX` to `IS_POSIX` (public constant) in `fab_secure_io.py` and updated the import in `fab_logger.py`.

File: tests/test_core/test_fab_logger.py [hunk 31]
```diff
@@ -249,3 +249,94 @@ def mock_log_warning():
...
+    # Write something to ensure the file exists
+    log_instance.debug("permission test entry")
+
+    assert log_file.exists()
+    mode = oct(log_file.stat().st_mode & 0o777)
+    assert mode == "0o600", f"Log file has mode {mode}, expected 0o600"
+
+    # Verify the owner can still read the file
+    content = log_file.read_text()
+    assert "permission test entry" in content
```
[the remark this answers is not in the document]
Comment (iemejia) [reply tests/test_core/test_fab_logger.py:299]: Fixed. Added handler cleanup at the end of the test: iterates `log_instance.handlers`, removes, and closes each handler to prevent accumulation on the global `logging.getLogger("FabricCLI")` singleton.

File: src/fabric_cli/core/fab_logger.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (aviatco) [reply src/fabric_cli/core/fab_logger.py]: Maybe you can move this logic to the helper you created, then you can avoid importing both _IS_POSIX  and chmod_if_posix

File: src/fabric_cli/core/fab_state_config.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (aviatco) [reply src/fabric_cli/core/fab_state_config.py]: Please close all the resolved comments you have in this PR

**Review by iemejia — COMMENTED** [review]
~Moved to individual thread replies below.~

**Review by iemejia — COMMENTED** [review]
~Duplicate, see above.~

File: src/fabric_cli/core/fab_logger.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (iemejia) [reply src/fabric_cli/core/fab_logger.py]: Done. Moved the restricted file opener logic into `fab_secure_io.py` as `get_restricted_file_opener()`, which returns the POSIX opener or `None` on Windows. `fab_logger.py` no longer imports `IS_POSIX` -- it only imports `chmod_if_posix`, `create_restricted_dir`, and `get_restricted_file_opener` from the helper.

File: src/fabric_cli/utils/fab_secure_io.py [hunk 32]
```diff
@@ -0,0 +1,70 @@
...
+
+    Logs at debug level when chmod fails (e.g. permission denied on
+    restrictive filesystems) since the user cannot act on it.
+    """
+    if IS_POSIX:
+        try:
+            os.chmod(path, mode)
+        except OSError as e:
+            _logger.debug("Failed to set permissions %o on %s: %s", mode, path, e)
+
```
[the remark this answers is not in the document]
Comment (iemejia) [reply src/fabric_cli/utils/fab_secure_io.py:34]: Keeping DEBUG intentionally. This was explicitly requested by @aviatco in an earlier review round (see resolved thread on `fab_file_permissions.py`): "Consider change warning to debug since in case of permission denied user has nothing to do with it. so, this is more debug error." The rationale is that chmod failures are not actionable by the user, and the primary protection comes from `os.open(..., 0o600)` at creation time -- the chmod is defense-in-depth for upgrading pre-existing files.

File: src/fabric_cli/core/fab_state_config.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (iemejia) [reply src/fabric_cli/core/fab_state_config.py]: Done. All previously addressed comment threads have been resolved.

File: pyproject.toml [hunk 33]
```diff
@@ -76,6 +76,11 @@ ignore_errors = true
 module = "pyspark.*"
 ignore_missing_imports = true
 
+[tool.black]
```
Comment (aviatco) [inline pyproject.toml]: please revert this change and keep it in tox.toml

Comment (iemejia) [reply pyproject.toml]: Done. Reverted the commit — `[tool.black]` config is back in `tox.toml`.

**Review by aviatco — APPROVED** [review]
